File: Core/Playlist/Playlists.cs
namespace Core.Playlist;

public static class Playlists
{
	public static int? IndexOfSong(PlaylistEntry[] entries, string songId)
	{
		var index = Array.FindIndex(entries, entry => entry.Id == songId);
		return index < 0 ? null : index;
	}

	public static PlaylistEntry[] Add(PlaylistEntry[] entries, PlaylistEntry entry) => [..entries, entry];

	public static PlaylistEntry[] RemoveAt(PlaylistEntry[] entries, int index)
	{
		if (index < 0 || index >= entries.Length)
			return entries;

		var list = entries.ToList();
		list.RemoveAt(index);
		return list.ToArray();
	}

	public static PlaylistEntry[] Move(PlaylistEntry[] entries, int oldIndex, int newIndex)
	{
		if (oldIndex < 0 || oldIndex >= entries.Length)
			return entries;

		var list = entries.ToList();
		var entry = list[oldIndex];
		list.RemoveAt(oldIndex);
		list.Insert(Math.Clamp(newIndex, 0, list.Count), entry);
		return list.ToArray();
	}

	public static PlaylistEntry[] Shuffle(PlaylistEntry[] entries, int? pinnedIndex, Random random)
	{
		var order = Enumerable.Range(0, entries.Length).ToArray();
		var start = 0;
		if (pinnedIndex is { } pinned)
		{
			(order[0], order[pinned]) = (order[pinned], order[0]);
			start = 1;
		}

		for (var i = order.Length - 1; i > start; i--)
		{
			var j = start + random.Next(i - start + 1);
			(order[i], order[j]) = (order[j], order[i]);
		}

		return order.Select(index => entries[index]).ToArray();
	}
}

File: shared/core/playlists.ts
// Pure playlist ordering transforms — the TS mirror of C# `Core.Playlist.Playlists`.
// Holds no cursor, gain, or position; operates only on the ordered entry array.

import type { PlaylistEntry } from './types';

// A source of randomness in [0, 1), injected so `shuffle` is deterministic in tests.
export type Rng = () => number;

export function indexOfSong(entries: PlaylistEntry[], songId: string): number | null {
	const index = entries.findIndex((entry) => entry.id === songId);
	return index < 0 ? null : index;
}

export function add(entries: PlaylistEntry[], entry: PlaylistEntry): PlaylistEntry[] {
	return [...entries, entry];
}

export function removeAt(entries: PlaylistEntry[], index: number): PlaylistEntry[] {
	if (index < 0 || index >= entries.length) {
		return entries;
	}
	return [...entries.slice(0, index), ...entries.slice(index + 1)];
}

export function move(entries: PlaylistEntry[], oldIndex: number, newIndex: number): PlaylistEntry[] {
	if (oldIndex < 0 || oldIndex >= entries.length) {
		return entries;
	}
	const list = [...entries];
	const [entry] = list.splice(oldIndex, 1);
	list.splice(clamp(newIndex, 0, list.length), 0, entry);
	return list;
}

export function shuffle(
	entries: PlaylistEntry[],
	pinnedIndex: number | null,
	rng: Rng
): PlaylistEntry[] {
	const order = entries.map((_, index) => index);
	let start = 0;
	if (pinnedIndex !== null) {
		[order[0], order[pinnedIndex]] = [order[pinnedIndex], order[0]];
		start = 1;
	}
	for (let i = order.length - 1; i > start; i--) {
		const j = start + Math.floor(rng() * (i - start + 1));
		[order[i], order[j]] = [order[j], order[i]];
	}
	return order.map((index) => entries[index]);
}

function clamp(value: number, low: number, high: number): number {
	return Math.min(Math.max(value, low), high);
}
